proposals/0149-migrate-snapshot-epoch-stakes.md

```diff
@@ -0,0 +1,159 @@
+    ..
+}
+
+struct EpochStakes {
+    #[serde(with = "crate::stakes::serde_stakes_enum_compat")]
+    stakes: Arc<StakesEnum>,
+    total_stake: u64,
+    node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
+    epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
+}
+
+enum StakesEnum {
+    Accounts(Stakes<StakeAccount>),
+    Delegations(Stakes<Delegation>),
+}
+```
+
+When serializing `EpochStakes` in snapshots, all `StakesEnum` variants first map
+stake entry values to their `Delegation` value before serialization. The goal of
+this proposal is to migrate to a new epoch stakes field which maps stake entry
+values to their full `Stake` value before serialization so that
+`credits_observed` will be be included in the snapshot and available after
+snapshot deserialization.
+
+The proposed `new_epoch_stakes` bank snapshot field will instead serialize a map
+of epochs to `VersionedEpochStakes` structs which can be updated in the future
+to serialize different information if needed. This field will be appended to the
+end of the serialized bank snapshot:
+
+```rust
+struct (BankFieldsToDeserialize | BankFieldsToSerialize) {
+    ..
+    new_epoch_stakes: HashMap<Epoch, VersionedEpochStakes>,
+}
+
+enum VersionedEpochStakes {
+    Current {
+        stakes: Stakes<Stake>,
+        total_stake: u64,
+        node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
+        epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
+    },
+}
+```
+
+### Snapshot Update Rollout
+
+Handling snapshot format changes is always a delicate operation to coordinate
+given that old software releases will not be able to deserialize snapshots from
+new software releases properly. The rollout will require two phases:
+
+1. Introduce support for deserializing the migrated epoch stakes field 
+2. Enable serializing epoch stakes to the new field and phase out the old field
+
+During the first phase, validator software will be updated to attempt to
+deserialize the new epoch stakes field appended at the end of the bank snapshot.
+If the field doesn't exist, validators will continue using the old deserialized
+field. If the new epoch stakes field is deserialized successfully, the epoch
+stakes entries from this field will be merged with entries from the old field.
+
+During the second phase, validator software will be updated to start serializing
+epoch stakes entries to the new epoch stakes field. Note, however, that there
+are 3 different epoch stakes entry variants:
+
+1. Entries created during epoch boundaries which have full stake account data
+2. Entries deserialized from the old snapshot epoch stakes field which only have
+    stake delegation state.
+3. Entries deserialized from the new snapshot epoch stakes field which have full
+    stake state.
+
+Only variants 1 and 3 can be serialized into the new epoch stakes field so any
+variant 2 epoch stakes entries will continue being serialized into the old epoch
+stakes field.
```

Comment:
It might be useful to describe what cases could cause variant 2 after a validator updates to the second-phase software. Will this be the case until the next epoch boundary after update?

Reply:
Also, please make explicit that only one of the fields will be serialized into for any given snapshot.

Reply:
I will add more details to make this clearer but basically we will have a mixed bag of epoch stakes entries during the migration. The new field won't support serializing variant 2 entries so I was planning to put them in the old field still even if we have already started using the new field.